controllers/cinder_controller.go

```diff
@@ -845,6 +845,24 @@ func (r *CinderReconciler) transportURLCreateOrUpdate(
 }
 
 func (r *CinderReconciler) apiDeploymentCreateOrUpdate(ctx context.Context, instance *cinderv1beta1.Cinder) (*cinderv1beta1.CinderAPI, controllerutil.OperationResult, error) {
+
+	apiTemplate := instance.Spec.CinderAPI
```

Comment:
Two suggestions to consider:
1. Use the same API for creating all 4 of cinder's deployments, by passing in the service template as you do with volumeDeploymentCreateOrUpdate(). I know that the volume service requires you to pass in the service template because there will be separate calls for each backend. What you could do here is pass in the API service template (on L847) in lieu of the variable on L849.
2. I don't think it's necessary to embed the service's name in the variable name (or the volTemplate parameter name on L974). You could use something like "svcTemplate" so that the code blocks are similar in all 4 locations.

Reply:
> Two suggestions to consider:
> 
>     1. Use the same API for creating all 4 of cinder's deployments, by passing in the service template as you do with volumeDeploymentCreateOrUpdate(). I know that the volume service requires you to pass in the service template because there will be separate calls for each backend. What you could do here is pass in the API service template (on L847) in lieu of the variable on L849.
> 
>     2. I don't think it's necessary to embed the service's name in the variable name (or the volTemplate parameter name on L974). You could use something like "svcTemplate" so that the code blocks are similar in all 4 locations.

The `CinderServiceTemplate` represents a common struct containing a list of fields that can be customized
 by the user for a given resource (`API/Sched/Bak/Vol`), because those can be changed/customized in the
 CR for the specific service: for instance, when the `CR` is defined, I can pass a diff `Resources` or `Replicas`
value for each Cinder service (or any other field provided by the `CinderServiceTemplate`).
This is the reason why I resolve the variable within the function (we pass `instance` here, and the templates
are already an attribute of its `Spec`: passing an additional variable containing data that we already pass
seems redundant to me).
Also, I don't always use `API` in all the 4 functions: each of them get the template values with the form:

```
<service>Template := instance.Spec.<Service>
```

For the reason above I'd like to keep this variable within the function, unless we have strong reasons to resolve
it outside and pass it to the function.

Regarding \#2, volumes are defined in the form:
```
CinderVolumes map[string]CinderServiceTemplate
```

which means that the `name` we extract in the `for loop` is the `key` of this map, and it represents the instance
name we want to pass to the k8s `Deployment`: passing it to the `VolCreateOrUpdate` function looks good to
me as we want to process,  for each backend, the associated parameters, and return the `CinderVolume` instance):
it's used, later in the code (but within the same for loop) to compute the status (`ReadyCount`, etc ..) and change the `Cinder.Status` accordingly.
I think it would result confusing changing the current logic, embedding the name (key) within a struct which represents
the associated value.

Reply:
I think we're in complete agreement on what the code is doing (I agree with everything you wrote), and the issue is one of coding style. I typically defer to the person who wrote the code, so I'm OK with the code as is (though I'd like Gorka to see this discussion).

BTW, the reason why I suggested using "svcConfig" for the local variable name (and function parameter in the Volume code) is it helps highlight common blocks of code. And the reason why I spotted this is that Gorka and I hope to (eventually) dedup code wherever possible.

Reply:
> I think we're in complete agreement on what the code is doing (I agree with everything you wrote), and the issue is one of coding style. I typically defer to the person who wrote the code, so I'm OK with the code as is (though I'd like Gorka to see this discussion).

+1

> 
> BTW, the reason why I suggested using "svcConfig" for the local variable name (and function parameter in the Volume code) is it helps highlight common blocks of code. And the reason why I spotted this is that Gorka and I hope to (eventually) dedup code wherever possible.

I'm ok to avoid duplication as much as possible (it would avoid the issues of the previous PS), and I think we have a couple of options to do that (just not sure we might want to have a follow up change on that part).
1. Try using generics: provide a function that accepts an interface and build the `Cinder<Service>Spec`; this is feasible but I think it would result hard to read/maintain, other that the fact we might still need logic to distinguish in between the Service passed/returned for a few specific fields
2. Embed the common fields of the `Cinder<Service>Spec` in the `CinderServiceTemplate` struct which can be defined in the services Spec itself.

E.g.

 ```
type CinderAPISpec {
   ...
   ...
   // +kubebuilder:validation:Required
   // CinderServiceTemplate specifies the generic fields of the service
   CinderServiceTemplate `json:",inline"`
  ...
```

and in the controller we can have something like:

```
	cinderAPISpec := cinderv1beta1.CinderAPISpec{
		CinderServiceTemplate: cinderv1beta1.CinderServiceTemplate{
			ContainerImage:             svcTemplate.ContainerImage,
			Replicas:                   svcTemplate.Replicas,
			Debug:                      svcTemplate.Debug,
			CustomServiceConfig:        svcTemplate.CustomServiceConfig,
			DefaultConfigOverwrite:     svcTemplate.DefaultConfigOverwrite,
			CustomServiceConfigSecrets: svcTemplate.CustomServiceConfigSecrets,
			NetworkAttachments:         svcTemplate.NetworkAttachments,
		},
		ServiceUser:                instance.Spec.ServiceUser,
		DatabaseHostname:           instance.Status.DatabaseHostname,
		DatabaseUser:               instance.Spec.DatabaseUser,
		Secret:                     instance.Spec.Secret,
		PasswordSelectors:          instance.Spec.PasswordSelectors,
		ExtraMounts:                instance.Spec.ExtraMounts,
		TransportURLSecret:         instance.Status.TransportURLSecret,
	}
```

This means, in theory, that the struct we pass in the above can be computed by a generic function, which is valid for all the Cinder services: this would avoid a lot of duplication given that the code would become:

```
svcTemplate := r.getServiceTemplate(instance.Spec.CinderAPI)
cinderAPISpec := cinderv1beta1.CinderAPISpec{
		CinderServiceTemplate: svcTemplate,
		ServiceUser:                instance.Spec.ServiceUser,
		DatabaseHostname:           instance.Status.DatabaseHostname,
		DatabaseUser:               instance.Spec.DatabaseUser,
		Secret:                     instance.Spec.Secret,
		PasswordSelectors:          instance.Spec.PasswordSelectors,
		ExtraMounts:                instance.Spec.ExtraMounts,
		TransportURLSecret:         instance.Status.TransportURLSecret,
	}
```

and that function should be something like:

```
func (r *CinderReconciler) getServiceTemplate(instance cinderv1beta1.CinderSpec) cinderv1beta1.CinderServiceTemplate {
	switch req {
	case "API":
		return instance.CinderAPI
	case "Scheduler":
		return instance.CinderScheduler
        ...
        ...
```

However, I'd like to keep a change like this for a follow up patch, because it doesn't change much on the CRD side (which is the purpose of this patch) and can be reviewed easily because it would be smaller and focused on the optimization front.

Reply:
I'm not concerned about deduping in this patch, so I'm fine with this patch as is. I think you can leave the dedup concerns to Gorka and me.